syntax = "proto3";
package smartbox_msgs;

service TrackerController {
	rpc get_tracker_status (TrackerSystemStatusRequest) returns (TrackerSystemStatusResponse) {}
	rpc tracker_control (stream ControlRequest) returns (stream ControlResponse) {}
	rpc tracker_status (TrackerSystemStatusRequest) returns (stream TrackerSystemStatusResponse) {}
	//rpc request_control (RequestControlRequest) returns (RequestControlResponse) {}
	//rpc relinquish_control (RelinquishControlRequest) returns (RelinquishControlResponse) {}
	//rpc move_panel (MoveRequest) returns (MoveResponse) {}
	rpc stop (StopRequest) returns (StopResponse) {}
	rpc stow (StowRequest) returns (StowResponse) {}
}

enum DIRECTION {
	NORTH = 0;
	EAST = 1;
	SOUTH = 2;
	WEST = 3;
}

enum ChargeState {
	START = 0;
	NIGHT_CHECK = 1;
	DISCONNECT = 2;
	NIGHT = 3;
	FAULT = 4;
	BULK_CHARGE = 5;
	ABSORBTION = 6;
	FLOAT = 7;
	EQUALIZE = 8;
}

enum ControlRequestSuccessFlag {
		SUCCESS = 0;
		INSUFFICIENT_SECURITY_LEVEL = 1;
		FAILURE=2;
	}


message TrackerSystemStatusRequest {
	string control_id = 1;
}

message Position {
	double ns = 1;
	double ew = 2;
}

message MoveStatus {
	bool ns = 1;
	bool ew = 2;
}

message TrackerStatus {
	Position position = 1;
	Position angle = 2;
	MoveStatus move_status = 3;
	int64 controlling_authority = 4;
	string controlling_client = 5;
	bool is_controlling_client = 6;
}	

message ChargeControllerDetails {
	double t_hs = 1;
	double t_batt = 2;
	double t_amb = 3;
	double t_rts = 4;

	uint32 array_fault = 5;
	double vb_f = 6;
	double vb_ref = 7;

	double ahc_r = 8;
	double ahc_t = 9;
	double kwhc = 10;
	double load_state = 11;
	uint32 load_fault = 12;
	double v_lvd = 13;

	double ahl_r = 14;
	double ahl_t = 15;

	double hourmeter = 16;
	uint32 alarm = 17;

	uint32 dip_switch = 18;
	uint32 led_state = 19;
	double power_out = 20;

	double sweep_vmp = 21;
	double sweep_pmax = 22;
	double sweep_voc = 23;

	double vb_min_daily = 24;
	double vb_max_daily = 25;

	double ahc_daily = 26;
	double ahl_daily = 27;

	uint32 array_fault_daily = 28;
	uint32 load_fault_daily = 29;

	uint32 alarm_daily = 30;
	double vb_min = 31;
	double vb_max = 32;
	uint32 lighting_should_be_on = 33;

	double va_ref_fixed = 34;
	double va_ref_fixed_ptc = 35;
}

message ChargeControllerStatus {
	double battery_voltage = 1;
	double array_voltage = 2;
	double load_voltage = 3;
	double charge_current = 4;
	double load_current = 5;

	ChargeState charge_state = 6;
	double energy_collected = 7;
	double energy_expended = 8;

	ChargeControllerDetails details = 9;
}

message RequestControlRequest {
	string description = 1;
	uint64 security_level = 2;
}

message RequestControlResponse {
	string message = 1;
	ControlRequestSuccessFlag success = 2;
	string control_id = 3;
}

message RelinquishControlRequest {
	string control_id = 1;
}

message RelinquishControlResponse {
	string message = 1;
	double energy_collected = 2;
	double energy_expended = 3;
	ControlRequestSuccessFlag success = 4;
}

message ControlRequest {
	enum MoveType {
		UNSET = 0;
		DURATION = 1;
		POSITION = 2;
		ANGLE = 3;
	}
	uint64 authority_level = 1;
	string description = 2;

	MoveType move_type = 3;
	DIRECTION direction = 4;
	double move_duration = 5;
	Position position = 6;
	Position angle = 7;
}

message ControlResponse {
	string message = 1;
	ControlRequestSuccessFlag success = 2;
}

message StopRequest {
	uint64 authority_level = 1;
	string description = 2;
}

message StopResponse {
	string message = 1;
	ControlRequestSuccessFlag success = 2;
}

message StowRequest {
	uint64 authority_level = 1;
	string description = 2;
}

message StowResponse {
	string message = 1;
	ControlRequestSuccessFlag success = 2;
}

message TrackerSystemStatusResponse {
	TrackerStatus tracker = 1;
	ChargeControllerStatus charge_controller = 2;	
}